hbwmalloc.3
.\"
.\" Copyright (C) 2014 Intel Corporation.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\" 1. Redistributions of source code must retain the above copyright notice(s),
.\"    this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright notice(s),
.\"    this list of conditions and the following disclaimer in the documentation
.\"    and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) ``AS IS'' AND ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
.\" EVENT SHALL THE COPYRIGHT HOLDER(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\" ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.TH "HBWMALLOC" 3 "2014-09-22" "Intel Corporation" "HBWMALLOC" \" -*- nroff -*-
.SH "NAME"
hbwmalloc \- The high bandwidth memory interface
.SH "SYNOPSIS"
.nf
.B #include <hbwmalloc.h>
.sp
.B Link with -ljemalloc -lnuma -lmemkind -lpthread
.sp
.B int hbw_check_available(void);
.br
.BI "void* hbw_malloc(size_t " "size" );
.br
.BI "void* hbw_calloc(size_t " "nmemb" ", size_t " "size" );
.br
.BI "void* hbw_realloc (void " "*ptr" ", size_t " "size" );
.br
.BI "void hbw_free(void " "*ptr" );
.br
.BI "int hbw_posix_memalign(void " "**memptr" ", size_t " "alignment" ", size_t " "size" );
.br
.BI "int hbw_posix_memalign_psize(void " "**memptr" ", size_t " "alignment" ", size_t " "size" ", int " "pagesize" );
.br
.B int hbw_get_policy(void);
.br
.BI "void hbw_set_policy(int " "mode" );
.fi
.SH "DESCRIPTION"
.BR hbw_check_available ()
returns 0 if high bandwidth memory is available and an error code
described in the
.B ERRORS
section if not.
.PP
.BR hbw_malloc ()
allocates
.I size
bytes of uninitialized high bandwidth memory. The allocated space is
suitably aligned (after possible pointer coercion) for storage of any
type of object. If
.I size
is zero then
.BR hbw_malloc ()
returns NULL.
.PP
.BR hbw_calloc ()
allocates space for
.I nmemb
objects in high bandwidth memory, each
.I size
bytes in length. The result is identical to calling
.BR hbw_malloc ()
with an argument of
.IR nmemb * size
, with the exception that the allocated memory is explicitly
initialized to zero bytes.  If
.I nmemb
or
.I size
is 0, then
.BR hbw_calloc ()
returns NULL.
.PP
.BR hbw_realloc ()
changes the size of the previously allocated high bandwidth memory
referenced by
.I ptr
to
.I size
bytes. The contents of the memory are unchanged up to the lesser of
the new and old sizes. If the new size is larger, the contents of the
newly allocated portion of the memory are undefined. Upon success, the
memory referenced by
.I ptr
is freed and a pointer to the newly allocated high bandwidth memory is
returned.

.BR Note:
.BR hbw_realloc ()
may move the memory allocation, resulting in a different return value
than
.IR "ptr" .

If
.I ptr
is NULL, the
.BR hbw_realloc ()
function behaves identically to
.BR hbw_malloc ()
for the specified size.
The address
.IR "ptr" ,
if not NULL, was returned by a previous call to
.BR hbw_malloc (),
.BR hbw_calloc (),
.BR hbw_realloc (),
.BR hbw_posix_memalign (),
or
.BR hbw_posix_memalign_psize ().
Otherwise, or if
.I hbw_free(ptr)
was called before, undefined behavior occurs.
.PP
.BR hbw_free ()
causes the allocated high bandwidth memory referenced by
.I ptr
to be made available for future allocations. If
.I ptr
is NULL, no action occurs.
The address
.IR "ptr" ,
if not NULL, must have been returned by a previous call to
.BR hbw_malloc (),
.BR hbw_calloc (),
.BR hbw_realloc (),
.BR hbw_posix_memalign (),
or
.BR hbw_posix_memalign_psize ().
Otherwise, if
.I hbw_free(ptr)
was called before, undefined behavior occurs.
.PP
.BR hbw_posix_memalign ()
allocates
.I size
bytes of high bandwidth memory such that the allocation's base address
is an even multiple of
.IR "alignment" ,
and returns the allocation in the value pointed to by
.IR "memptr" .
The requested
.I alignment
must be a power of 2 at least as large as
.IR "sizeof(void *)" .
.PP
.BR hbw_posix_memalign_psize ()
allocates
.I size
bytes of high bandwidth memory such that the allocation's base address
is an even multiple of
.IR "alignment" ,
and returns the allocation in the value pointed to by
.IR "memptr" .
The requested
.I alignment
must be a power of 2 at least as large as
.IR "sizeof(void *)" .
The memory will be allocated using pages determined by the
.IR "psize"
variable which may be one of the following enumerated values:
.TP
.B HBW_PAGESIZE_4KB
The four kilobyte page size option. Note that with transparent huge
pages enabled these allocations may be promoted by the operating
system to two megabyte pages.
.TP
.B HBW_PAGESIZE_2MB
The two megabyte page size option.
.TP
.B HBW_PAGESIZE_1GB
This option allows the user to specify arbitrary sizes backed by
one gigabytes pages.
.TP
.B HBW_PAGESIZE_1GB_STRICT
The one gigabyte page size option. The total size of the allocation
must be a multiple of 1GB with this option, otherwise the allocation
will fail.
.PP
.BR hbw_get_policy ()
returns the current fallback policy when insufficient high bandwith
memory is available.
.PP
.BR hbw_set_policy ()
set the current fallback policy, the policy can be modified only once
in the life of an application. The policy in effect at the time when a
buffer is allocated determines the policy for that buffer until it is
freed.
.TP
.B HBW_POLICY_BIND
If insufficient high bandwidth memory is available to satisfy
a request, the allocated pointer is set to NULL and
.I errno
is set to ENOMEM.
.TP
.B HBW_POLICY_PREFERRED
If insufficient high bandwidth memory is available, fall back to
standard memory (default).
.SH "RETURN VALUE"
.BR hbw_get_policy ()
returns
.B HBW_POLICY_BIND
or
.B HBW_POLICY_PREFERRED
which represents the current high bandwidth policy.
.BR hbw_set_policy ()
and
.BR hbw_free ()
do not have return values.
.BR hbw_malloc ()
.BR hbw_calloc ()
and
.BR hbw_realloc ()
return the pointer to the allocated memory, or NULL if the request
fails.
.BR hbw_posix_memalign ()
and
.BR hbw_posix_memalign_psize ()
return zero on success and return an error code
as described in the
.B ERRORS
section below on failure.
.SH ERRORS
.TP
.BR hbw_check_available ()
returns one of the error codes described in
.BR memkind (3)
if high bandwidth memory is not available.
.TP
.BR "hbw_posix_memalign" "() and " "hbw_posix_memalign_psize" "()"
return the one of the POSIX standard error codes
.B EINVAL
or
.B ENOMEM
as defined in
.IR <errno.h>
if an error occurs.
If the
.I alignment
parameter is not a power of two, or was not a multiple of
.IR "sizoeof(void *)" ,
then
.B EINVAL
is returned.  If there was insufficient memory to satisfy the request then
.B ENOMEM
is returned.
.SH "NOTES"
The
.I hbwmalloc.h
file defines the external functions and enumerations for the hbwmalloc
library. These interfaces define a heap manager that targets high
bandwidth memory numa nodes.
.SH "FILES"
.TP
.I /etc/memkind/node-bandwidth
File that contains the bandwidth values for each numa node.
.TP
.I /etc/rc/d/init.d/memkind
Initialization script that creates the node-bandwidth file by calling
the PMTT table parser.
.TP
.I /usr/sbin/memkind-pmtt
The PMTT table parser.

.SH "ENVIRONMENT"
.TP
.B MEMKIND_HBW_NODES
This environment varaible is a comma separated list of NUMA nodes that
are treated as high bandwidth. This environment variable should be set
if the PMTT file is not present, or to override the PMTT table if it
is present. Uses the
.I libnuma
routine
.BR numa_parse_nodestring ()
for parsing, so the syntax described in the
.BR numa (3)
man page for this routine applies for example: 1-3,5 is a valid setting.
.SH "COPYRIGHT"
Copyright (C) 2014 Intel Corporation. All rights reserved.
.SH "SEE ALSO"
.BR malloc (3),
.BR numa (3),
.BR numactl (8),
.BR mbind (2),
.BR mmap (2),
.BR move_pages (2)
.BR jemalloc (3)
.BR memkind (3)


hbwmalloc.c
/*
 * Copyright (C) 2014 Intel Corperation.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 1. Redistributions of source code must retain the above copyright notice(s),
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice(s),
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
 * EVENT SHALL THE COPYRIGHT HOLDER(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <pthread.h>
#include <errno.h>

#include "hbwmalloc.h"
#include "memkind.h"

static int hbw_policy_g = HBW_POLICY_PREFERRED;
static pthread_once_t hbw_policy_once_g = PTHREAD_ONCE_INIT;
static inline memkind_t hbw_get_kind(int pagesize);
static inline void hbw_policy_preferred_init(void);
static inline void hbw_policy_bind_init(void);


int hbw_get_policy(void)
{
    return hbw_policy_g;
}

void hbw_set_policy(int mode)
{
    if (mode == HBW_POLICY_PREFERRED) {
        pthread_once(&hbw_policy_once_g, hbw_policy_preferred_init);
    }
    else if (mode == HBW_POLICY_BIND) {
        pthread_once(&hbw_policy_once_g, hbw_policy_bind_init);
    }
    if (mode != hbw_policy_g) {
        fprintf(stderr, "WARNING: hbw_set_policy() called more than once with different values, only first call heeded.\n");
    }
}

int hbw_check_available(void)
{
    return memkind_check_available(MEMKIND_HBW);
}

void *hbw_malloc(size_t size)
{
    memkind_t kind;

    kind = hbw_get_kind(HBW_PAGESIZE_4KB);
    return memkind_malloc(kind, size);
}

void *hbw_calloc(size_t num, size_t size)
{
    memkind_t kind;

    kind = hbw_get_kind(HBW_PAGESIZE_4KB);
    return memkind_calloc(kind, num, size);
}

int hbw_posix_memalign(void **memptr, size_t alignment, size_t size)
{
    memkind_t kind;

    kind = hbw_get_kind(HBW_PAGESIZE_4KB);
    return memkind_posix_memalign(kind, memptr, alignment, size);
}

int hbw_posix_memalign_psize(void **memptr, size_t alignment, size_t size,
                                int pagesize)
{
    memkind_t kind;
    int err = 0;

    kind = hbw_get_kind(pagesize);
    if (pagesize == HBW_PAGESIZE_1GB_STRICT &&
        size % 1073741824) {
        err = EINVAL;
    }
    if (!err) {
        err = memkind_posix_memalign(kind, memptr, alignment, size);
    }
    return err;
}

void *hbw_realloc(void *ptr, size_t size)
{
    memkind_t kind;

    kind = hbw_get_kind(HBW_PAGESIZE_4KB);
    return memkind_realloc(kind, ptr, size);
}

void hbw_free(void *ptr)
{
    memkind_t kind;
    memkind_get_kind_for_free(ptr, &kind);
    memkind_free(kind, ptr);
}

static inline memkind_t hbw_get_kind(int pagesize)
{
    memkind_t result = NULL;

    if (hbw_get_policy() == HBW_POLICY_BIND) {
        switch (pagesize) {
            case HBW_PAGESIZE_2MB:
                result = MEMKIND_HBW_HUGETLB;
                break;
            case HBW_PAGESIZE_1GB:
            case HBW_PAGESIZE_1GB_STRICT:
                result = MEMKIND_HBW_GBTLB;
                break;
            default:
                result = MEMKIND_HBW;
                break;
        }
    }
    else if (memkind_check_available(MEMKIND_HBW) == 0) {
        switch (pagesize) {
            case HBW_PAGESIZE_2MB:
                result = MEMKIND_HBW_PREFERRED_HUGETLB;
                break;
            case HBW_PAGESIZE_1GB:
            case HBW_PAGESIZE_1GB_STRICT:
                result = MEMKIND_HBW_PREFERRED_GBTLB;
                break;
            default:
                result = MEMKIND_HBW_PREFERRED;
                break;
        }
    }
    else {
        switch (pagesize) {
            case HBW_PAGESIZE_2MB:
                result = MEMKIND_HUGETLB;
                break;
            case HBW_PAGESIZE_1GB:
            case HBW_PAGESIZE_1GB_STRICT:
                result = MEMKIND_GBTLB;
                break;
            default:
                result = MEMKIND_DEFAULT;
                break;
        }
    }
    return result;
}


static inline void hbw_policy_bind_init(void)
{
    hbw_policy_g = HBW_POLICY_BIND;
}

static inline void hbw_policy_preferred_init(void)
{
    hbw_policy_g = HBW_POLICY_PREFERRED;
}
